### PR TITLE
Cascade removal of client sessions on parent user session removal

### LINKED ISSUE #12096: HotRod user session store doesn't remove client sessions on user session removal
### Describe the bug

When a user session is removed we want to have all client sessions for this user removed. This is done using [UserSessionConcurrentHashMapStorage](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/storage/chm/UserSessionConcurrentHashMapStorage.java)  for CHM, however, we don't have anything like this for HotRod storage.

NOTE: this could be solved by invalidate method, however, this is probably not suitable because we want to allow each storage to implement this cascade removal on its own.  

### Version

18.0.0

### Expected behavior

Client sessions are removed on user session removal

### Actual behavior

Client sessions are not removed

### How to Reproduce?

1. create a user session
2. create a client session for ^
3. remove user session
4. client session should be removed but it is not for HotRod storage

### Anything else?

_No response_

### LINKED ISSUE #12146: `UserSessionConcurrentHashMapStorage` does not remove client sessions on user session removal
### Describe the bug

The reason seems to be that `clientSessionStore` within `UserSessionConcurrentHashMapStorage` is not enlisted to `KeycloakTransactionManager`. We need to make sure `UserSessionConcurrentHashMapStorage` use the same client session transaction as is used in `MapUserSessionProvider`.

### Version

18.0.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mhajas**
Closes https://github.com/keycloak/keycloak/issues/12146
Closes https://github.com/keycloak/keycloak/issues/12096

### CODE REVIEW

**Comment by mhajas** [comment]
Currently, I am not able to say whether the failure in Model Tests is caused by these changes. However, I tried to run more iterations (6) and all of them passed: https://github.com/mhajas/keycloak/actions/runs/2390100196

**Review by martin-kanis — CHANGES_REQUESTED** [review]
Thank @mhajas for the PR. I have one comment.

File: model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProviderFactory.java [hunk 1]
```diff
@@ -289,19 +297,31 @@ public String getCacheName() {
...
+        if (modelType == UserSessionModel.class) getHotRodStorage(session, AuthenticatedClientSessionModel.class, flags);
+        return storages.computeIfAbsent(modelType, c -> createHotRodStorage(session, modelType, flags));
+    }
+
+    private <E extends AbstractHotRodEntity, V extends HotRodEntityDelegate<E> & AbstractEntity, M> HotRodMapStorage<String, E, V, M> createHotRodStorage(KeycloakSession session, Class<M> modelType, MapStorageProviderFactory.Flag... flags) {
+        HotRodConnectionProvider connectionProvider = session.getProvider(HotRodConnectionProvider.class);
+
+        if (modelType == UserSessionModel.class) {
+            HotRodMapStorage clientSessionStore = getHotRodStorage(session, AuthenticatedClientSessionModel.class);
+            HotRodEntityDescriptor<HotRodUserSessionEntity, HotRodUserSessionEntityDelegate> entityDescriptor = (HotRodEntityDescriptor<HotRodUserSessionEntity, HotRodUserSessionEntityDelegate>) getEntityDescriptor(modelType);
```
[the remark this answers is not in the document]
Comment (mhajas) [reply model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorageProviderFactory.java]: Fixed

**Review by vramik — APPROVED** [review]
I have checked non-HotRod part of the PR and it LGTM, thanks @mhajas

**Review by martin-kanis — APPROVED** [review]

**Review by hmlnarik — APPROVED** [review]
Approving per @vramik and @martin-kanis 's approvals

**Pull request merged**